Dev (#1)
wise-vision/project-cezary: README.md

# Soil moisture system using grove capicty moisture sensor, Wio E5 dev board and Wio E5 mini board
## Introduction

> ![System schema](./docs/asset/project_cezary_1.drawio.png)

Reviewer: Add addional line after because new line is appended to image
![image](https://github.com/wise-vision/project-cezary/assets/20014205/4b1e735d-a646-4e9f-8e06-cb0edc72d243)

Reviewer 2: Also, use the same grammatical structure for all of the bullet points. e.g.:
The soil moisture system includes:
- Reading the soil moisture level via ADC on the mini board.
- Transmitting data about the moisture level via LoRa to the dev board.
- Receiving this data by the dev board and sending instructions to the mini board about which color the RGB LED should light up.

Additionally, in the image above, I assume that the left board is the mini board, but it is not 100% clear. You can add some explanation at the very beginning, e.g., "The system consists of the following hardware: ..." 
Or you can just combine all of these information into one list :smiley:

Soil moisture system includes:
* Reading soil mositure level via ADC on mini board.

> * Sending data about mositure level via Lora to dev board.

Reviewer: Nit:
```suggestion
* Sending data about mositure level via LoRa to dev board.
```

> * Dev board recive this data and send order to mini board about, what color rgb led, should lights.

Reviewer: ```suggestion
* Dev board receive this data and send order to mini board about, what color rgb led, should lights.
```

## Technologies
Project is created with:
* Zephyr: 3.5.0
* West: 1.2.0

> ## Preparation

Reviewer 2: Prerequisites

1. It is necessary to have Zephyr already installed to build or flash application.

>    * Go to the official Zephyr RTOS website (https://www.zephyrproject.org/) to download the latest version.

Reviewer: I would add more direct link to download page. When I opened this link, there is no "Download" button visible.

>    * Follow the installation instructions on the website to install Zephyr RTOS on your system.

Reviewer 2: If we redirect somebody to another website, I would also give a link, so that you don't have to look for the right instructions.

2. Install Build Tools:

> * Zephyr RTOS uses build tools like CMake and ninja. You have to follow this commmands to install:

Reviewer 2: I don't like the phrase "You have to follow this commands".
I would change the whole part to:

Zephyr RTOS uses build tools like CMake and Ninja:
```
sudo apt-get update
sudo apt-get install cmake
sudo apt-get install ninja-build
```

If you want to you can change the first line to something like:
Zephyr RTOS uses build tools like CMake and Ninja. Use following commands to install both:

* CMake:
```
sudo apt-get update
sudo apt-get install cmake
```
* Ninja:
```

> sudo apt-get update

Reviewer 2: As this is one preparation process, you don't have to add `sudo apt-get update` for each build tool. I would leave the first one only.

sudo apt-get install ninja-build
```
3. Install picom
* To communicate with your board you have to install picom:
```
sudo apt-get update
sudo apt-get -y install picom
```                                                         
## E5 dev board setup             

> To build versions for the e5 dev board, you need an unlocked board(this shows how to unlock board https://wisevision.tech/docs/LoRa/LoRa-e5-dev-board-unprotect-memory), connected via UART and st-link.

Reviewer 2: What are "versions" here?

Space before opening the brackets, no comma after brackets.

"connected via UART and st-link."
I am not an expert in the field, but is it obvious to what the board should be connected?  PC in this case?

> 1. You need to copy Source code from github into zephyrpoject directory:

Reviewer 2: Don't start every instruction with "you need". Simple: "Copy the source code from github into `/zephyrpoject` directory" is enough.

```
cd /zephyrpoject
git clone https://github.com/wise-vision/project-cezary.git
```

> 2. You need to set the path to the code file:

Reviewer 2: `cd` is used to change directory. 
Maybe something like:
Navigate to the build directory:
or
Go the the build directory:

Change the "build directory" to whatever this directory refers to :smiley:

```
cd /zephyrpoject/project-cezary/boards/e5_dev_board
```
3. Build source code:
```
west build -b lora_e5_dev_board
```
Successful build should looks like this:
![Successful build](./docs/correct_build.png)
4. Upload your build to board:
```
west flash
```
Successful flash should looks like this:

> ![Successful flash](./docs/asset/correct_flash.png)

Reviewer: Maybe it might be good to add some information about connecting board to computer i.e. if some programmer/debugger is needed or just standard USB cable (and what type: C, micro, mini etc). Probably update graphics from the beginning.

> ![Successful flash](./docs/asset/correct_flash.png)

Reviewer 2: I don't like the idea of attaching screenshots of terminal output.
I can see that the first image corresponds to build process. Maybe text description like this can replace the image:
Build is successful when all of the build jobs are finished. To test, upload the build to the board.

Regarding the second image, I think @adamkrawczyk is a much more experienced person to review.

> 5. After build and flash to the board, to run the app you have to press the reset button on board.

Reviewer 2: remove "you have to"

![Wio E5 dev board reset button](./docs/asset/reset_dev_board.png)
## E5 mini board setup

> To build versions for the e5 mini board, you need an unlocked board(this shows how to unlock board https://wisevision.tech/docs/LoRa/LoRa-e5-dev-board-unprotect-memory), connected via UART and st-link.

Reviewer 2: ditto

> 1. You need to copy Source code from github into zephyrpoject directory:

Reviewer 2: ditto for whole section

```
cd /zephyrpoject
git clone https://github.com/wise-vision/project-cezary.git
```
2. You need to set the path to the code file:
```
cd /zephyrpoject/project-cezary/boards/e5_mini
```
3. Build source code:
```
west build -b lora_e5_dev_board
```
Successful build should looks like this:
![Successful build](./docs/correct_build.png)
4. Upload your build to board:
```
west flash
```
Successful flash should looks like this:
![Successful flash](./docs/asset/correct_flash.png)
5. After build and flash to the board, to run the app you have to press the reset button on board.
![Wio E5 mni board reset button](./docs/asset/reset_mini_board.png)

> ## Code of E5 dev board

Reviewer 2: For the whole section, please refactor. First shortly explain what the piece of code does, next place the piece of code, and next add some more details about the code if needed.

### prj.conf
```c
CONFIG_LOG=y
CONFIG_SPI=y
CONFIG_GPIO=y
CONFIG_LORA=y
CONFIG_CBPRINTF_FP_SUPPORT=y
```

> These configurations help customize the Zephyr OS for specific hardware and functionality requirements.

Reviewer 2: Place explanation before the code.

### src/main.c
```c
int configureRx(const struct device *const lora_dev)
{
	struct lora_modem_config config;
	config.frequency = 865100000;
	config.bandwidth = BW_125_KHZ;
	config.datarate = SF_10;
	config.preamble_len = 8;
	config.coding_rate = CR_4_5;
	config.iq_inverted = false;
	config.public_network = false;
	config.tx_power = 14;
	config.tx = false;

	int ret = lora_config(lora_dev, &config);
	if (ret < 0) {
		LOG_ERR("LoRa config failed");
	}
	return ret;
}
```

> This function is for configuarting reception. Lora newtork has few things to confing, but we have only to choose the frequnecy. Frequnecy is diffrent for diffrent regions.

Reviewer: ```suggestion
This function is for receiving configuration. LoRa network has a few things to configure, but we have to choose only the frequnecy. Frequnecy is different for different regions.
```

Reviewer 2: ditto

Freuqency plans:
* Europe: 863-870 Mhz
* United States: 902-928 Mhz
* Austraulia: 915-928 Mhz
* China 470-510 Mhz
* Asia/Japan 920-923 Mhz
* Morocco 869-870 Mhz
* India 865-867 Mhz
* Russia 864-870 Mhz

```c
int configureTx(const struct device *const lora_dev)
{
	struct lora_modem_config config;
	config.frequency = 865100000;
	config.bandwidth = BW_125_KHZ;
	config.datarate = SF_10;
	config.preamble_len = 8;
	config.coding_rate = CR_4_5;
	config.iq_inverted = false;
	config.public_network = false;
	config.tx_power = 4;
	config.tx = true;

	int ret = lora_config(lora_dev, &config);
	if (ret < 0) {
		LOG_ERR("LoRa config failed");
	}
	return ret;
}
```

> This function is for configuarting transmission. Like in previous function we have to choose to correct frequnecy.

Reviewer: ```suggestion
This function is for transmission configuration. Like in previous function, we have to choose to correct frequnecy.
```

```c
void recev_msg(const struct device *dev, uint8_t *data, int16_t rssi, int8_t snr){
	int len;
	configureRx(dev);
	LOG_INF("Recive data");
	len = lora_recv(dev, data, MAX_DATA_LEN, K_FOREVER, &rssi, &snr);
	if (len < 0) {
		LOG_ERR("LoRa receive failed");
		return 0;
	}

	LOG_INF("Received data: %s (RSSI:%ddBm, SNR:%ddBm)", data, rssi, snr);
}
```
This function reciving data from min board and shows info about moisture in `LOG_INF`
```c
void send_msg(const struct device *dev, uint8_t *data, int16_t rssi, int8_t snr){
	char data_send[MAX_DATA_LEN] = {0};
	int ret;
	LOG_INF("Sending data");
	int converted_value = atoi(data);
	if (converted_value<30){
	data_send[0]='1';
	}else{
	data_send[0]='0';
	}
	configureTx(dev);
	ret=lora_send(dev,data_send,1);
	if (ret < 0) {
		LOG_ERR("LoRa send failed");
	}else{
		LOG_INF("Data send:%s",data_send);
	}

}
```

> This function send data 0 if the soil moisture is accurate and 1 if our plant needs water. The correct level of moisture is set for under 30 %. It could be change for your purpose.

Reviewer: ```suggestion
This function send data 0 if the soil moisture is correct and 1 if our plant needs water. The correct level of moisture is set for under 30 %. It could be change for your purpose.
```
I assume that you meant correct in this case i.e. soil moisture is at the correct level.

```c
int main(void)
{
	const struct device *const lora_dev = DEVICE_DT_GET(DEFAULT_RADIO_NODE);
	uint8_t data[MAX_DATA_LEN] = {0};
	int16_t rssi;
	int8_t snr;

	if (!device_is_ready(lora_dev)) {
		LOG_ERR("%s Device not ready", lora_dev->name);
		return 0;
	}
	while(1){
		recev_msg(lora_dev, data, rssi, snr);
		send_msg(lora_dev, data, rssi, snr);
		k_sleep(K_MSEC(1000));
	}
	k_sleep(K_FOREVER);
	return 0;
}
```

> In main function we recev and send messegs form board to bord with 1 second interval.

Reviewer: ```suggestion
In main function we receive and send messegs form board to board with 1 second interval.
```

```c

/ {
	zephyr,user {
		/* adjust channel number according to pinmux in board.dts */
		io-channels = <&adc1 2>;
	};
};

&adc1 {
	#address-cells = <1>;
	#size-cells = <0>;
	pinctrl-0 = <&adc_in2_pb3>;
	
	status = "okay";
	
		channel@2 {
		reg = <2>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <12>;
		};
};
```
This part of device tree defines Analog to Digital converter pins. Channel 2 is used because in E5 mini board it is deafult channel.
### prj.conf
```c
CONFIG_ADC=y
CONFIG_SHELL=y
CONFIG_LOG=y
CONFIG_SPI=y
CONFIG_GPIO=y
CONFIG_LORA=y
```
It has the same use like in dev board, but its added more configs:
*ADC for reading values from moisture sensor.
*GPIO for changeing colour of led.
### src/main.c
`configureRx and configureTx`
This functions are the same as on dev boards. If we want communication between two boards, we have to set the same frequency.
```c
const int DryValue = 1950;
const int WetValue = 1220;
int mapSoilMoisture(int value, int inMin, int inMax, int outMin, int outMax) {
    return (value - inMin) * (outMax - outMin) / (inMax - inMin) + outMin;
}
```

> This function is used to translet masurments from ADC to 0-100 values. If we want to have correct values of moisture, we have to make a calibration of the sensor. 

Reviewer: ```suggestion
This function is used to translate measurement from ADC to 0-100 values. If we want to have correct values of moisture, we have to calibrate the sensor. 
```

#### Calibration of the sensor
* Check the value of dry sensor and write it into DryValue
* Check the value of the sensor puted in the water and write it into WetValue.
```c
void send_msg(const struct device *dev, int moisture){
	int ret;
	LOG_INF("Received moisture value: %d\n", moisture);
	char data[MAX_DATA_LEN];
	char moisture_str[MAX_DATA_LEN];
	snprintf(moisture_str, sizeof(moisture_str), "%.2d", moisture);
	memset(data, 0, sizeof(data));
	strncpy(data, moisture_str, sizeof(data) - 1);
	configureTx(dev);
	ret = lora_send(dev, data, MAX_DATA_LEN);
	if (ret < 0) {
		LOG_ERR("LoRa send failed");
		return 0;
	}

	LOG_INF("Data sent!");

}
```
This function sends data too dev board about soil moisture.
```c
void recev_msg(const struct device *dev, uint8_t *data_recive, int16_t rssi, int8_t snr){
	int len;
	configureRx(dev);
	LOG_INF("Recive data");
	len = lora_recv(dev, data_recive, MAX_DATA_LEN, K_FOREVER, &rssi, &snr);
	if (len < 0) {
		LOG_ERR("LoRa receive failed");
		return 0;
	}

	LOG_INF("Received data: %s", data_recive);
}
```

> This function recevs data from dev board.

Reviewer: ```suggestion
This function receives data from dev board.
```

```c
void gpios_config(const struct device *gpio1, const struct device *gpio2){
	int ret;
	ret = gpio_pin_configure(gpio1, 0, GPIO_OUTPUT_ACTIVE);//green
	if (ret != 0) {
		return;
	}
		ret = gpio_pin_configure(gpio2, 10, GPIO_OUTPUT_ACTIVE);//green
	if (ret != 0) {
		return;
	}
	if(!device_is_ready(gpio1)){
		return;
	}
	if(!device_is_ready(gpio2)){
		return;
	}

}
```

> This function configs gpios outputs for led controling.

Reviewer: ```suggestion
This function configures gpios outputs for led controlling.
```

```c
void adc_config(){
	int err;
	for (size_t i = 0U; i < ARRAY_SIZE(adc_channels); i++) {
		if (!adc_is_ready_dt(&adc_channels[i])) {
			LOG_INF("ADC controller device %s not ready\n", adc_channels[i].dev->name);
			return 0;
		}

		err = adc_channel_setup_dt(&adc_channels[i]);
		if (err < 0) {
			LOG_INF("Could not setup channel #%d (%d)\n", i, err);
			return 0;
		}
	}
}
```

> This function is used to configs analog to digital converter.

Reviewer: ```suggestion
This function is used to configure analog to digital converter.
```

		.buffer = &buf,
		.buffer_size = sizeof(buf),
	};
	int err;
	int moisture;
	for (size_t i = 0U; i < ARRAY_SIZE(adc_channels); i++) {
		int32_t val_mv;
		(void)adc_sequence_init_dt(&adc_channels[i], &sequence);
		err = adc_read_dt(&adc_channels[i], &sequence);
		if (err < 0) {
			
			continue;
		}
		if (adc_channels[i].channel_cfg.differential) {
			val_mv = (int32_t)((int16_t)buf);
		} else {
			val_mv = (int32_t)buf;
		}
		err = adc_raw_to_millivolts_dt(&adc_channels[i],&val_mv);
		int val_int = (int)val_mv;
		moisture = mapSoilMoisture(val_int, DryValue, WetValue, 0, 100);

	}
	return moisture;
}
```
This function returns values of moisture in range 0-100.
```c
void set_led(uint8_t *data_recive, const struct device *gpio1, const struct device *gpio2){
	int ret;
	int SetLed = atoi(data_recive);
	if (SetLed == 1) {
		
		ret = gpio_pin_set_raw(gpio2, 10, 0);
		ret = gpio_pin_set_raw(gpio1, 0, 1);	
			
	} else {
		
		ret = gpio_pin_configure(gpio1, 5,GPIO_OUTPUT_ACTIVE); //red
		if (ret != 0) {
			return;
		}
		
		ret = gpio_pin_set_raw(gpio1, 0, 0);
		ret = gpio_pin_set_raw(gpio2, 10, 1);
			
	} 
}
```
This function sets color of leds depending on the values send from dev board.

> ## Finishing

Reviewer 2: Add newline before each section

> ## Finishing

Reviewer 2: I don't like "Finishing" as the name of the section but I don't have any idea how to name it. @styczen @adamkrawczyk

Reviewer 3: Maybe `Conclusion` ?

> If you flashed your board with built programms you can check their logs.

Reviewer 2: programs

To see logs
1. Open terminal
2. Use command `sudo picocom --baud 115200 /dev/"your_usb_port"`
### On Wio E5 dev board kit logs should looks like this:
![Wio E5 dev board logs](./docs/asset/logs_dev_board.png)

> * Recive data informs us that data has been recived

Reviewer: ```suggestion
* Receive data informs us that data has been received
```

> * Recived data shows value of moisture

Reviewer: ```suggestion
* Received data shows value of moisture
```

> * Sending data infroms us that data has been sended

Reviewer: ```suggestion
* Sending data informs us that data has been sended
```

* Data send shows value that has been send: 1 means you need too water the plant and 0 means that your soil moisture is correct
* Blinking led on board informs us that there is connection via uart.
### On Wio E5 mini board logs should looks like this:

> ![Wio E5 dev board logs](./docs/asset/logs_mini_board.png)

Reviewer 2: This is why I don't like pictures of text in the README. The images are not proportional and the size of letters varies a lot. Use:
```sh
blabla 
goo
foo
```

Reviewer 3: Add extra line here

> * Data send! infroms us that data has been sended

Reviewer 2: Try to be more careful with grammar :) Also, when you can, try to avoid making the instruction personal. In most cases you don't need to use words like: "us", "you", "your".
"Data Sent!" informs that the data has been sent.
"Receive Data" informs that the data has been received.
"Receive Data" shows the value sent from the dev board: 1 indicates a need for watering the plant, and 0 indicates that the soil moisture is correct.

Reviewer 2: Change the whole section accordingly

* Recive data informs us that data has been recived

> * Recive data shows value sended from dev board: 1 means you need too water the plant and 0 means that your soil moisture is correct

Reviewer: ```suggestion
* Receive data informs us that data has been received
* Receive data shows value sended from dev board: 1 means you need too water the plant and 0 means that your soil moisture is correct
```
